src/utils/static_vector.hh

```diff
@@ -30,45 +33,78 @@ template<typename T, size_t N>
+
 	constexpr static_vector() = default;
 
 	explicit constexpr static_vector(std::initializer_list<T> list) {
 		assert(list.size() <= N);
-		std::copy(list.begin(), list.end(), data.data());
+		ranges::copy(list, array);
 		sz = SizeType(list.size());
 	}
 
 	template<typename Range>
 	constexpr static_vector(from_range_t, Range&& range) {
 		for (auto&& elem : range) {
-			push_back(elem);
+			push_back(std::forward<decltype(elem)>(elem));
 		}
 	}
 
-	[[nodiscard]] constexpr auto begin() const noexcept { return data.begin(); }
-	[[nodiscard]] constexpr auto end()   const noexcept { return data.begin() + sz; }
-
-	[[nodiscard]] constexpr size_t size() const { return sz; }
-	[[nodiscard]] constexpr bool empty() const { return sz == 0; }
+	[[nodiscard]] constexpr auto begin()        noexcept { return array.begin(); }
+	[[nodiscard]] constexpr auto begin()  const noexcept { return array.begin(); }
+	[[nodiscard]] constexpr auto end()          noexcept { return array.begin() + sz; }
+	[[nodiscard]] constexpr auto end()    const noexcept { return array.begin() + sz; }
+	[[nodiscard]] constexpr auto rbegin()       noexcept { return       reverse_iterator(end()); }
+	[[nodiscard]] constexpr auto rbegin() const noexcept { return const_reverse_iterator(end()); }
+	[[nodiscard]] constexpr auto rend()         noexcept { return       reverse_iterator(begin()); }
+	[[nodiscard]] constexpr auto rend()   const noexcept { return const_reverse_iterator(begin()); }
 
-	[[nodiscard]] constexpr       T& operator[](size_t index)       { return data[index]; }
-	[[nodiscard]] constexpr const T& operator[](size_t index) const { return data[index]; }
+	[[nodiscard]] constexpr size_t size()     const noexcept { return sz; }
+	[[nodiscard]] constexpr size_t max_size() const noexcept { return array.max_size(); }
+	[[nodiscard]] constexpr bool   empty()    const noexcept { return sz == 0; }
 
-	constexpr void push_back(const T& a) { assert(sz < N); data[sz++] = a; }
+	[[nodiscard]] constexpr       T& operator[](size_t index)       noexcept { return array[index]; }
+	[[nodiscard]] constexpr const T& operator[](size_t index) const noexcept { return array[index]; }
+	[[nodiscard]] constexpr T&       front()       noexcept { assert(sz > 0); return array[0]; }
+	[[nodiscard]] constexpr const T& front() const noexcept { assert(sz > 0); return array[0]; }
+	[[nodiscard]] constexpr T&       back()        noexcept { assert(sz > 0); return array[sz - 1]; }
+	[[nodiscard]] constexpr const T& back()  const noexcept { assert(sz > 0); return array[sz - 1]; }
+	[[nodiscard]] constexpr       T* data()        noexcept { return array.data(); }
+	[[nodiscard]] constexpr const T* data()  const noexcept { return array.data(); }
 
-	constexpr void clear() { sz = 0; }
+	template<typename... Args>
+	constexpr T& emplace_back(Args&&... args) {
+		assert(sz < N);
+		std::construct_at(std::addressof(array[sz]), std::forward<Args>(args)...);
+		++sz;
+		return back();
+	}
```

Comment:
This is wrong in general (but works in our current use case).

`std::construct_at()` constructs an object in _uninitialized_ memory (memory where no other object is currently living). So it does NOT first destruct the object at the given address.

We should do this destruct manually via something like `array[sz].~T()`.
In our current use case in AmdFlash the type in `static_array` has a trivial destructor (=destructor does nothing), so this doesn't matter.

More in general: the current implementation stores the data in an `std::array<T, N>`. But that means that for non-trivial types at construction we construct `N` times  a `T` object. And similar in the destructor. A better `static_vector` implementation could be to not use a `std::array` but just an array of uninitialized memory (with proper size and alignment) and then manually manage the lifetime of all stored objects.

We could try that, or maybe simpler for now only support types that have a trivial destructor. So maybe add something like: `static_assert(is_trivially_destructible_v<T>);`  (Given that we don't currently need this and that in the future we want to switch to `std::inplace_vector` which does handle this all correctly).

Reply:
Another option is to get rid of `emplace()` I guess? But it’d be nicer to have it, and if `clear()` would destruct, etc.

I’ll see what I can do.

Reply:
I'm not sure what you mean with "_...and if `clear()` would destruct ..._". In the current version it's wrong for `clear()` to destruct, because then later when `static_vector` itself gets destroyed, the elements get destroyed twice. So I assume you meant proper lifetime management for all operations (and this implies we can no longer use a `std::array<T, N>` as the internal storage).

I see 3 possible solutions:
* Do a minimum local fix in emplace_back(). This could be adding a call to `~T()`. Or it could be: first construct a temporary object and then move-assign that over the existing element in the array.
* Document and enforce the limitation that our `static_vector` is only meant to be used with trivially destructible types. So add that `static_assert` I mentioned above.
* Implement proper lifetime management. Maybe we can get inspiration from this [reference implementation](https://godbolt.org/z/5P78aG5xE). Unfortunately we can't use this implementation directly because it uses some c++23 features (but no doubt it's possible to adapt this).

The last option would be technically the most interesting (a good learning experience), BUT at the moment we don't need the extra functionality, and the intention remains to eventually replace our implementation with a standard version (c++26?).

Personally I'd go for the 2nd option. But I'd be happy to help you with any other option as well.

Reply:
What I mean by “_...and if `clear()` would destruct ..._” is if only the actual elements in use were constructed, so constructing the vector would not default-construct, push / emplace would construct, and pop / clear would destruct.

> I see 3 possible solutions:

The fourth possible solution is to remove `emplace_back()`, right? Like in the new change I just pushed.

This fourth option would be my choice atm as the most trivial and safe choice. This commit in the pull request would then just add some functionality and not (fail to) attempt to optimise the construction. Optimizing the construction and lifetime can be done separately, later.

Even with a `static_assert` that restricts it to trivial types I don’t necessarily have full confidence that it would do the correct thing in all possible situations. I feel there are nuances that I don’t necessarily have the deep C++ language knowledge for to properly anticipate. I lost some confidence here ;). So at the moment it feels safer to me to steer clear of memory and lifetime management entirely, I don’t need it to achieve my goals with this PR.

Is this ok with you?

Reply:
All these options are fine for me. I did a very quick review of the latest version of this PR, and I'm OK with pulling it as it is now.  Do you permission to do this? (I think you have.) Or should I do it?

A fifth alternative could be to implement `emplace_back()` in terms of `push_back()`. That way users of `static_vector` can already write `emplace_back()`, but it only becomes an advantage in the future, e.g. when we switch to `std::inplace_vector`.
Maybe something like this (untested):
```
template<typename... Args>
T& emplace_back(Args&& ...args) {
    push_back(T(std::forward<Args>(args)...));
    return back();
}
```
But feel free to ignore this suggestion.